### PR TITLE
Adding gates to the MPS simulator

### PR DESCRIPTION
**Pull request by yaelbh**
### Summary
Addressing #1008

### Details and comments
`cy` is done

### CODE REVIEW

**Comment by yaelbh** [comment]
`csx` is done

**Comment by yaelbh** [comment]
`delay` is done

**Comment by yaelbh** [comment]
`cswap` is done

**Comment by yaelbh** [comment]
`diagonal` is done

**Comment by yaelbh** [comment]
Why is the job on Windows cancelled?

File: test/terra/reference/ref_1q_clifford.py [hunk 1]
```diff
@@ -776,3 +776,124 @@ def pauli_gate_counts_deterministic(shots, hex_counts=True):
         targets.append({'101': shots})
         targets.append({'011': shots})
     return targets
+
+# ==========================================================================
+# I-gate
+# ==========================================================================
+
+def id_gate_circuits_deterministic(final_measure=True):
```
Comment (chriseclectic) [inline test/terra/reference/ref_1q_clifford.py:784]: What is the purpose for adding these test circuits? Isn't it enough to have a test for identity and delay in the standard gate tests?

File: test/terra/backends/qasm_simulator/qasm_cliffords.py [hunk 2]
```diff
@@ -255,7 +255,53 @@ def test_pauli_gate_deterministic_default_basis_gates(self):
...
-                      **self.BACKEND_OPTS)
+                      **self.BACKEND_OPTS, optimization_level=0)
+        result = job.result()
+        self.assertSuccess(result)
+        self.compare_counts(result, circuits, targets, delta=0)
+
+    # ---------------------------------------------------------------------
+    # Test identity gate
+    # ---------------------------------------------------------------------
+    def test_id_gate_deterministic(self):
```
Comment (chriseclectic) [inline test/terra/backends/qasm_simulator/qasm_cliffords.py:266]: Arent these tests getting removed in #1068 anyway?

**Review by chriseclectic — COMMENTED** [review]
I think we should merge #1068 before this PR to avoid adding unnecessary tests

File: src/simulators/matrix_product_state/matrix_product_state_tensor.hpp [hunk 3]
```diff
@@ -320,52 +344,109 @@ void MPS_Tensor::apply_tdg()
...
+  }
+  else { 
+    indices.push_back(2);
+    indices.push_back(3);
+  }
+  
+  apply_matrix_helper(mat, is_diagonal, indices);
+}    
+
+void MPS_Tensor::apply_matrix_helper(const cmatrix_t &mat, bool is_diagonal,
```
Comment (merav-aharoni) [inline src/simulators/matrix_product_state/matrix_product_state_tensor.hpp:394]: Nice, I think you improved `apply_matrix`.

File: src/simulators/matrix_product_state/matrix_product_state_tensor.hpp [hunk 4]
```diff
@@ -320,52 +344,109 @@ void MPS_Tensor::apply_tdg()
...
-  if (swapped)
-    swap(data_[1], data_[2]);
+    for (uint_t i=0; i<mat.GetRows(); i++)
+      data_[indices[i]] = new_data[i];
+  }
 }
 
 void MPS_Tensor::apply_cnot(bool swapped)
 {
   if(!swapped)
```
Comment (merav-aharoni) [inline src/simulators/matrix_product_state/matrix_product_state_tensor.hpp:421]: Be consistent with your usage of `if (swapped)` or `if (!swapped)`. In two places you use the former and here you use the latter.

**Review by merav-aharoni — APPROVED** [review]

File: test/terra/reference/ref_1q_clifford.py (hunk 1 above)
Comment (yaelbh) [reply test/terra/reference/ref_1q_clifford.py:784]: The purpose here is to test the stabilizer simulator for the identity and delay gates

File: test/terra/backends/qasm_simulator/qasm_cliffords.py (hunk 2 above)
Comment (yaelbh) [reply test/terra/backends/qasm_simulator/qasm_cliffords.py:266]: Not for the stabilizer simulator

File: src/simulators/matrix_product_state/matrix_product_state_internal.cpp [hunk 5]
```diff
@@ -556,25 +572,34 @@ void MPS::apply_3_qubit_gate(const reg_t &qubits,
...
+    sub_tensor.apply_ccx(target);
+    break;
+  case cswap:
+    uint_t control;
+    if (qubits[0] < qubits[1] && qubits[0] < qubits[2])
+      target = 0;
+    else if (qubits[0] > qubits[1] && qubits[0] > qubits[2])
+      target = 2;
+    else
+      target = 1;
```
[the remark this answers is not in the document]
Comment (yaelbh) [reply src/simulators/matrix_product_state/matrix_product_state_internal.cpp:601]: thanks!

**Comment by yaelbh** [comment]
Fixed according to review comments

**Comment by yaelbh** [comment]
@mtreinish no conflicts now, can you please take a look at the transpiler's error. Thanks.

**Comment by yaelbh** [comment]
@mtreinish @chriseclectic  Do you understand the failure on Windows?  Looks like it abruptly stopped running in the middle of testing.

**Comment by mtreinish** [comment]
> @mtreinish @chriseclectic Do you understand the failure on Windows? Looks like it abruptly stopped running in the middle of testing.

Yes, that's exactly what happend. It looks like the test worker is segfaulting or some other system level crash which is causing the test worker from exiting without reporting the status back to the parent test runner process.

https://github.com/Qiskit/qiskit-aer/pull/1088/checks?check_run_id=1757586259#step:8:2997
```
{1} test.terra.backends.test_qasm_simulator_matrix_product_state.TestQasmMatrixProductStateSimulator.test_diagonal_gate [] ... inprogress
```

means that the test worker number 1 sent an event that it was starting that test, but never reported back the test had finished or what the result of the test was before the worker process exited. In almost all cases this means the worker crashed for some reason, and since there is no other output associated with the failure it's likely a segfault or some other system level failure that happens outside of python's control (so it can't report it).

**Comment by yaelbh** [comment]
@mtreinish thanks for the quick answer!  I'll look for memory corruption for the diagonal gate in MPS.

**Comment by yaelbh** [comment]
CI is passing, ready for a new review.  @chriseclectic @mtreinish

File: qiskit/providers/aer/backends/qasm_simulator.py [hunk 6]
```diff
@@ -482,12 +482,12 @@ def _method_configuration(method=None):
         # Matrix product state method
         elif method == 'matrix_product_state':
             config.description = 'A C++ QasmQobj matrix product state simulator with noise'
-            config.custom_instructions = sorted(['roerror', 'snapshot', 'kraus'])
```
Comment (chriseclectic) [inline qiskit/providers/aer/backends/qasm_simulator.py:485]: This needs fixing after #1109

File: qiskit/providers/aer/backends/qasm_simulator.py [hunk 7]
```diff
@@ -482,12 +482,12 @@ def _method_configuration(method=None):
...
-            config.basis_gates = sorted([
-                'u1', 'u2', 'u3', 'u', 'p', 'cp', 'cx', 'cz', 'id', 'x', 'y', 'z', 'h', 's',
-                'sdg', 'sx', 't', 'tdg', 'swap', 'ccx', 'unitary', 'delay',
-                'r', 'rx', 'ry', 'rz', 'rxx', 'ryy', 'rzz', 'rzx', 'initialize',
-            ] + config.custom_instructions)
```
Comment (chriseclectic) [inline qiskit/providers/aer/backends/qasm_simulator.py:490]: Same here

**Review by chriseclectic — CHANGES_REQUESTED** [review]

File: qiskit/providers/aer/backends/qasm_simulator.py (hunk 7 above)
Comment (yaelbh) [reply qiskit/providers/aer/backends/qasm_simulator.py:490]: Thanks!  done

**Review by chriseclectic — APPROVED** [review]

**Pull request merged**